lib/engines/hcktest/hcktest.rb

```diff
@@ -183,31 +184,25 @@ def configure_setup_and_synchronize
       @client1 = @clients.values[0]
       @client2 = @clients.values[1]
       @client1.support = @client2
+      @studio.keep_snapshot
+      @clients.each_value(&:keep_snapshot)
     end
 
-    def clean_last_run_clients
-      @clients.values.map(&:clean_last_run)
-    end
-
-    def clean_last_run_machines
-      @studio.clean_last_run
-      clean_last_run_clients
-    end
-
-    def run_and_configure_setup
+    def run_and_configure_setup(scope)
       retries ||= 0
 
-      run_studio
-      sleep 5 until @studio.up?
-      run_clients keep_alive: true
+      ResourceScope.open do |tmp_scope|
+        run_studio tmp_scope
+        sleep 5 until @studio.up?
+        run_clients tmp_scope, keep_alive: true
 
-      configure_setup_and_synchronize
+        configure_setup_and_synchronize
+        tmp_scope.move_to scope
+      end
     rescue AutoHCKError => e
-      synchronize_clients(exit: true)
       @project.logger.warn("Running and configuring setup failed: (#{e.class}) #{e.message}")
       raise e unless (retries += 1) < AUTOHCK_RETRIES
 
-      clean_last_run_machines
```

Comment:
Will we recreate snapshots when configuring setup failed? Sorry, this is not clear to me.

Reply:
Yes, when a setup fails, it will first close the current resource scope, which makes snapshots deleted. Then, if the retry counter is not saturated, it will rerun from the start, creating new snapshots.